### PR TITLE
:sparkles: envtest utility, defaultable flags

### PR DESCRIPTION
**Pull request by DirectXMan12**
This adds a small example/utility that lets you run envtest as a standalone binary.   I've probably written this code about 20 times in `/tmp/xyz-debug` when debugging controller issues or reproducing apiserver behavior, so I figured it was time to push it somewhere.

This also tweaks the way a flag gets initialized in the zap flags so that the value passed to options gets set as the default for `--zap-devel`.  This should make that flag behave like the other flags -- if you pass a "default" via the options struct and the user doesn't set the flag, they'll get that default.

### CODE REVIEW

**Comment by DirectXMan12** [comment]
@vincepri or someone, mind taking a quick peek at this

/assign @vincepri

File: examples/scratch-env/go.mod [hunk 1]
```diff
@@ -0,0 +1,11 @@
+module sigs.k8s.io/controller-runtime/examples/scratch-env
+
+go 1.14
```
Comment (vincepri) [inline examples/scratch-env/go.mod:3]:
```suggestion
go 1.15
```

File: examples/scratch-env/go.mod [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+replace sigs.k8s.io/controller-runtime => ../..
```
Comment (vincepri) [inline examples/scratch-env/go.mod:5]: Nit: Usually I've seen `replace` directives go at the very bottom

File: examples/scratch-env/main.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+			}
+			if amt != len(contents) {
+				log.Error(io.ErrShortWrite, "unable to write all of the kubeconfig file")
+				return true
+			}
+
+			return false
+		}(); exit {
+			return 1
+		}
```
Comment (vincepri) [inline examples/scratch-env/main.go:100]: Do we want to split this into a different function?

File: examples/scratch-env/main.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+		log.Info("wrote kubeconfig")
```
Comment (vincepri) [inline examples/scratch-env/main.go:102]:
```suggestion
		log.Info("Wrote kubeconfig")
```

File: examples/scratch-env/main.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+	}
+
+	<-ctrl.SetupSignalHandler()
+
+	log.Info("shutting down apiserver & etcd")
```
Comment (vincepri) [inline examples/scratch-env/main.go:107]:
```suggestion
	log.Info("Shutting down apiserver & etcd")
```

File: examples/scratch-env/main.go [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+	err = env.Stop()
+	if err != nil {
+		log.Error(err, "unable to stop the test environment")
+		return 1
+	}
+
+	log.Info("shutdown succesful")
```
Comment (vincepri) [inline examples/scratch-env/main.go:114]:
```suggestion
	log.Info("Shutdown succesful")
```

File: examples/scratch-env/main.go [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+	// TODO(directxman12): add support for writing to a new context in an existing file
+	kubeconfigFile, err := ioutil.TempFile("", "scratch-env-kubeconfig-")
+	if err != nil {
+		log.Error(err, "unable to create kubeconfig file, continuing on without it")
+	} else {
+		defer os.Remove(kubeconfigFile.Name())
+
+		log := log.WithValues("path", kubeconfigFile.Name())
+		log.V(1).Info("writing kubeconfig")
```
Comment (vincepri) [inline examples/scratch-env/main.go:64]:
```suggestion
		log.V(1).Info("Writing kubeconfig")
```

File: examples/scratch-env/main.go [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+	env := &envtest.Environment{}
+	env.CRDInstallOptions.Paths = *crdPaths
+	env.WebhookInstallOptions.Paths = *webhookPaths
+
+	if *attachControlPlaneOut {
+		os.Setenv("KUBEBUILDER_ATTACH_CONTROL_PLANE_OUTPUT", "true")
+	}
+
+	log.Info("starting apiserver & etcd")
```
Comment (vincepri) [inline examples/scratch-env/main.go:47]:
```suggestion
	log.Info("Starting apiserver & etcd")
```

**Review by vincepri — COMMENTED** [review]
Mostly nits, everything else lgtm

**Comment by DirectXMan12** [comment]
all fixed up

**Review by vincepri — APPROVED** [review]
/lgtm

**Pull request merged**